lucene/core/src/java/org/apache/lucene/codecs/VectorValues.java

```diff
@@ -0,0 +1,75 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.lucene.codecs;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.TopDocs;
+import org.apache.lucene.util.BytesRef;
+
+/**
+ * Holds the vector values for a particular field.
+ *
+ * @lucene.experimental
+ */
+public abstract class VectorValues {
+
+  public abstract DocIdSetIterator iterator();
+
+  public abstract int docID();
+
+  public abstract float[] value() throws IOException;
```

Comment:
Also, I think we should at least consider what our approach might be to other underlying numeric datatypes. Someday we may want to support low-fidelity vectors of bytes with a similar API. Would we then produce ByteVectorValues? I guess that might be best given generics can't model primitive types, in which case the only question is naming. Is it OK to privilege float vectors - I guess so, but felt it was at least worth considering

Reply:
> I'm concerned that we don't bake in unneccessary copying and/or allocation in the API

+1, this is an important point to consider. I guess another option would be to only expose certain functions like `double distance(float[] queryVector)`, but this seems quite restrictive.

> Also, I think we should at least consider what our approach might be to other underlying numeric datatypes.

I was also thinking we would have different classes here, for example `BitVectorValues` to handle bit vectors, etc. This would mirror the approach for doc values, and it also gives the flexibility to tailor the ANN approach to the specific vector type. To me it seems okay to rename this to `FloatVectorValues`, I don’t think it’s a big overhead.

Reply:
I think we can add more APIs for accessing  a doc's vector as we find them necessary.

I would assume that `float[] value()` should be called only a small number of docs during for example rescoring step. Is this correct assumption?

Reply:
@mayya-sharipova the hope was to provide a general API -- besides the ANN use cases we know about, I'm not sure we can anticipate exactly how people will build off of `VectorValues`. So I do think it makes sense to consider more streamlined access options.

Reply:
I'm not sure `void value(float[])` is very convenient since it require the caller to know the number of dimensions before making the call? Both the current API and the FloatsRef proposal would work for me however.

I'd like to avoid multiplying the number of these interfaces if we want to support more data types in the future. I'd rather let the format detect whether not all bits of the floats are used and optimize the case when all values are 0/1 (bits) or integers in the [0,256] range accordingly. This is how doc values work today by adapting the storage layout to the actual data. Another option might be to use byte[] and make the interpretation of these bytes an implementation detail of the codec, but this also makes the API a bit harder to work with.

Reply:
I copied this patch locally and tweaked it in order to answer some questions that were gnawing at me about performance. I don't think in the end that they are significant for the API being defined here, but have more to do with its implementation, and seem to require API changes elsewhere (in IndexInput), but I'll share here for posterity.

I was exploring the idea that we could get improved performance by computing dot-products of batches of vectors rather than fetch/decode/compute one at a time. I found a modest improvement from doing that (20%) but it isn't significant enough that I think we should expose a batch fetch method in this API.

A much larger improvement resulted from being able to decode floats directly from the internal buffer managed by IndexInput (qua ByteBufferIndexInput). I had exposed its internal buffer when testing the batch computation, and this allows for wrapping the ByteBuffer.asFloatBuffer(), avoiding the copy we do in the example implementation in this CR. This led to about a 2x speedup in a microbenchmark computing lots of dot-products.